raiden/network/proxies/secret_registry.py

```diff
@@ -61,13 +66,24 @@ def __init__(
         self._open_secret_transactions_lock = Semaphore()
 
     def register_secret(self, secret: Secret, given_block_identifier: BlockSpecification):
-        self.register_secret_batch([secret], given_block_identifier)
+        try:
+            self.register_secret_batch([secret], given_block_identifier)
+        except NoStateForBlockIdentifier:
+            # If the given block identifier is older than 128 blocks then try again
+            # with the latest block. This can happen if we shut down for some time
+            # and we get to register_secret when replaying the state changes.
+            self.register_secret_batch([secret], given_block_identifier)
```

Comment:
This is trying again with the same block.

Additionally, I don't this is a valid solution, IMO we should just skip the `precondition` checks, having two calls is just wasteful.

Reply:
> So if the given block is too old you want to skip the preconditions checks the second time and go ahead and send the transaction anyway

No. Instead of having two calls to `register_secret_batch` I would have only one, no `try-except` here. The behavior of the `register_secret_batch` would be:

- Check if for the given block number the state can be queried, if it cannot then we skip the preconditions checks. (This only means we can't detect programming errors).
- Then do the `eth_estimateGas`  call to see if the transaction can be sent, the logic would proceed as it is, if the gas estimation fails, the transaction would fail, and therefore it's not sent.

The point here is: The `given_block_identifier` is the block identifier which generate the request, and it's sole purpose is to allows us to do precondition checking, and the sole purpose of the precondition checking is to detect programming errors. If we change the `given_block_number` to `latest`, then we are using the wrong generating block, and the precondition checks will have a false positive, e.g.:

Sending a secret register *after the secret has registered and the node has processed the given block* is a programming error, so consider this case:

- The secret is registered a block `n`
- The node emits a `ContractSendSecretRegister` at block `n + x`
- Because the node's block is larger then `n`, we know that the node should have processed that block and it must know the secret was registered, therefore we have a programming error.

Consider the following case:

- The node emits a `ContractSendSecretRegister` at block `n` and crashes
- The secret is registered by some other node at a block `n + x`
- The node restarts at block `n + x + y`, at this point, the precondition check should not fail, but the transaction should not be sent, because at the time the node tried to register the secret, that action was valid, but at the time the secret was being registered, it was not necessary anymore

Now, if `y` from above is large enough, say `128` block, and we change `n` to  `latest`, then what would be a perfectly fine call becomes a programming error. For this reason, I don't think we should change that value.

Reply:
Check what I wrote in rocketchat. I agree in principle, except that in this case we can't rely on estimate gas. But good point. Instead of going from the beginning again, just do the secret is registered query on the latest hash in the except block.